# Notas Criptografía

## 8 abril

1. Logaritmo discreto: 
 a, b, p primo, encontrar x tq $a^x = b \mod p$. Fuerza bruta $log_b \mod p$
2. Factorización
3. Raices modulares

In [15]:
import practica1 as p1
import math

### Paso enano paso gigante
1. Cojemos un primo p fuerte de 3 cifras 263
2. Cojemos un generador, $a^{p-1} \neq 1$ a = 112
3. Calcular $112^{115}\mod 263 = 182$ b= 182
4. ¿como encontrar 115? multiplicar a*a hasta 115

1. Calculamos la raiz cuadrada de 263 = 16,....
2. Tomamos s = 17 (justo por encima de la raiz)
3. Tabla: [$182, 182 \times 112, 182 \times 112^2, .. 122^{17}]$
4. Empezamos en la tabla $112^{17}, 112^{17\times 2}....112^{17 \times 17}$
5. [141, 156, 167, 140, 15, 11, 236]
6. ¿Cual esta en la lista anterior? 236 si esta
7. Una vez que encontramos la coincidencia con el anterior: $117^{17 \times 7}=182 \times 112^4$ Despejando 112: $182 = 112^{115}$

In [18]:
def primera_tabla(p,a,gen):
    tabla = []
    s = int(math.sqrt(p)+0.5)
    exp = gen
    for i in range (1,s):
        a = (a*exp)%p
        tabla.append(a)
    return tabla

def segunda_tabla(primera, p, a, gen, it):    
    s = int(math.sqrt(p)+0.5)
    a = p1.potencia_modular(gen,s,p)
    exp = s
    tabla [gen]
    for i in range (1,s+1):
        a = p1.potencia_modular(gen,exp,p)
        exp = g
        
        
print (primera_tabla(263,182,112))
    

[133, 168, 143, 236, 132, 56, 223, 254, 44, 194, 162, 260, 190, 240, 54]


Cojemos un numero x, y= x^2,x + algo y ahora al numero sqrt(y) por encima y deberia de salir el numero x+1. 

In [19]:
x = 1002020202020202002020020202020032032030203203020302
y= (x*x)

La función sqrt redondea y no nos sirve 
### Metodo de newton ranson
Como un sgd pero no busca el minimo si no que el punto de la recta X^2 es 0(valor de la raiz)

In [38]:
def raiz(n):
    # Contar cuantas veces calcula la tangente
    i=0
    m = len(bin(n))//2
    x = 2**m
    y = ((x**2)+n)//(2*x)
    
    while x > y:
        x=y
        y=(x**2 + n) // (2*x)
        i+=1
        
    return (i,x)
x = 48987987654465132540
y = x*x +1
print (math.sqrt(y))
print (raiz (y))

4.8987987654465135e+19
(5, 48987987654465132540)


### P de Pollard
1. Quiero calcular $log_a(b) \mod p$. 
2. [x_1,x_2....x_n] sucesión aleatoria en $Z_p$ y encontrar una coincidencia.
3. Sucesion x1, x2 , x3=x20



## 13 abril
### RSA
p,q dos primos, 
n = p*q
mcd(e,fi(n))
(n,e) clave publica, d clave privada (e*d=1 mod fi(n))
#### Exponentes de cifrado 
Paginas 162..
##### Problemas de RSA
1. Factorizar n
2. Encontrar fi(n)
3. encontrar mcm(p-1,q-1)(lambda (n))
Los tres problemas equivalen a factorizar n 
#### Ejercicio probar  a romper usando fi(n)= 135645218856, n= 135651523807, p+q = 6304952
p*q=n 
fi(n)=(p-1)*(q-1)= pq-q-p+1
p+q = n-fi(n)+1

a*b=84 // (19-b)b=84->b^2 -19b+84=0
a+b=19 -> a = 19-b
solucion p =21589, q= 6283363 resolver p²-(s*p)+n=0 (s = p+q)


In [19]:
import factorizacion as fc
import util.math as mt
import util.primality as pr
from random import randint
f_n = 135645218856
n = 135651523807
p_q = 6304952


seg = mt.square((p_q**2)-(4*n))
print (seg)
print ("(b+seg)/2a = {}".format((p_q+seg)/2))
print ("(b+seg)/2a = {}".format((p_q-seg)/2))

6261774
(b+seg)/2a = 6283363.0
(b+seg)/2a = 21589.0


### Datos 
n es producto de p y q
lmbda(m) = mcm(p-1,q-1)

#### ejemplo
n=589
lambda(n)=90
1. calcular r = n mod lambda(n), r= 49
2. Dos posibilidades: 
n-fi(n) = r //-> fi(n)= n-r
n-fi(n) = r + lambda(n) //-> fi(n)=n-r-lambda(n)

In [17]:
n=589
lm=90
r=49
fi_n1 = n-r
fi_n2 = n-r-lm
p_q_1 = n-fi_n1+1
p_q_2 = n-fi_n2+1

seg = mt.square((p_q_1**2)-(4*n))
print ("(b+seg)/2a = {}".format((p_q_1+seg)/2))
print ("(b+seg)/2a = {}".format((p_q_1-seg)/2))
seg = mt.square((p_q_2**2)-(4*n))
print ("(b+seg)/2a = {}".format((p_q_2+seg)/2))
print ("(b-seg)/2a = {}".format((p_q_2-seg)/2))

(b+seg)/2a = 31.0
(b+seg)/2a = 19.0
(b+seg)/2a = 135.5
(b+seg)/2a = 4.5


n= 135651523807
e = 65537
clave publica (n,e) (hay 6 exponentes de cifrado) uno de ellos es d= 12920070665
1. Comprobamos que es un exponente de cifrado
m(aleatorio) 
m^(e*d) mod n 
2. conociendo d podemos factorizar n
d*e -1 = 2^u * s


In [28]:
m = randint(1,n-1)
e = 65537
n= 135651523807
d1= 12920070665
print (m)
print (mt.power_optimize(m,e*d1,n))
d2=80742680093 # otro exponente de cifrado
print (mt.power_optimize(m,e*d2,n))

28511152834
28511152834
28511152834


In [56]:
def descomponer(a):
    c = 0
    b = a
    while b % 2 == 0:
        c = c + 1
        b = b // 2
    return c, b

# escribimos d*e-1
u,s = descomponer((d1*e)-1)
print ("u = {}, s = {}".format(u,s))
#x = randint (2,n-2)
#x = 81736467419
x = 96735889556
y = mt.power_optimize(x,s,n)
print ("x = {}, y= {}".format(x,y))
y_2 = (y**2)%n
y_4 = (y**4)%n
y_8 = (y**8)%n
# Cojemos 3 porque u = 3
print ("y_2 = {}, y_4 = {}, y_8 = {}".format(y_2,y_4,y_8))
print ("mcd (z-1,n)={}".format(mt.mcd(y_2-1,n)))

u = 3, s = 105842833896513
x = 96735889556, y= 134533085192
y_2 = 106100867619, y_4 = 1, y_8 = 1
mcd (z-1,n)=6283363


z^2 % n = 1
z^2 -1 es multiplo de n
(z+1)(z-1) es multiplo de n